Add a background job to prune outdated sync tokens

tcitworld opened the pull request:
We remove all outdated sync tokens, based on their auto-incremented ID.

By default we only keep the last 10 000, but this can be configurable.

Closes #3349

We can add
```php
$this->setTimeSensitivity(IJob::TIME_INSENSITIVE);
```
if #30945 is merged first.

miaulalala commented:
Looks good, but code scanning is complaining a bit :)

nickvergessen commented:
> if https://github.com/nextcloud/server/pull/30945 is merged first.

It's merged

ChristophWurst commented on apps/dav/lib/BackgroundJob/PruneOutdatedSyncTokensJob.php:
```diff
@@ -0,0 +1,70 @@
+ */
+namespace OCA\DAV\BackgroundJob;
+
+use OC\BackgroundJob\TimedJob;
```
take the one from OCP?

ChristophWurst commented on apps/dav/lib/BackgroundJob/PruneOutdatedSyncTokensJob.php:
```diff
@@ -0,0 +1,70 @@
+	public function run($argument) {
+		$limit = (int) $this->config->getAppValue(Application::APP_ID, 'totalNumberOfSyncTokensToKeep', '10000');
+
+		// 0 or less is more likely a misconfiguration from the app config than an actual value.
```
worth a debug log I'd say

tcitworld replied:
Used `max()` to make sure we always get a positive non-null int instead.

ChristophWurst commented on apps/dav/lib/CalDAV/CalDavBackend.php:
```diff
@@ -3161,6 +3161,17 @@ protected function getCalendarObjectId($calendarId, $uri, $calendarType):int {
 	}
 
+	/**
+	 * @throws \OCP\DB\Exception
```
I wouldn't even document this, because we don't expect an error to occur here during normal operation. the exception will only be seen when the database is down or similar, right?

tcitworld replied:
I usually put this everywhere IDE suggests me, so that it's kept in mind when using the method, and there's some other methods in this file in the same case. 

In this specific case there's at least possibility of an exception when `$keep` is `< 0`, but I should probably move the check here and throw an `\InvalidArgumentException`.

ChristophWurst commented on apps/dav/tests/unit/BackgroundJob/PruneOutdatedSyncTokensJobTest.php:
```diff
@@ -0,0 +1,106 @@
+			->method('getAppValue')
+			->with(Application::APP_ID, 'totalNumberOfSyncTokensToKeep', '10000')
+			->willReturn($configValue);
+
```
nitpick: it can be beneficial to structure tests by the AAA pattern (Arrange, Act, Assert), so that you have three blocks separated by an empty line each. In that sense I would remove the blank lines until the last mock setup (arrange) and then have one blank line before calling the job (act). As the test subject doesn't return anything there is no assert block.

tcitworld replied:
There's a lot of tests that have will have an unreadable Arrange part given the number of calls made if I remove all their blank lines. :)

But in this specific case, will do.

tcitworld commented on apps/dav/lib/CalDAV/CalDavBackend.php:
```diff
@@ -3088,6 +3088,20 @@ protected function getCalendarObjectId($calendarId, $uri, $calendarType):int {
+		}
+		$query = $this->db->getQueryBuilder();
+		$query->delete('calendarchanges')
+			->orderBy('id', 'DESC')
```
Should probably be `ASC` in fact. :see_no_evil:

tcitworld committed:
Add a background job to prune outdated sync tokens

We remove all outdated sync tokens, based on their auto-incremented ID.

By default we only keep the last 10 000, but this can be configurable.

Signed-off-by: Thomas Citharel <tcit@tcit.fr>

szaimen commented:
/rebase
